### PR TITLE
NSFS | NC | Manage NSFS | Create bucket - Add Check If Bucket Owner Has `rw_access` to Underlying Path

### LINKED ISSUE #7883: NSFS | NC | CLI | Create Bucket - add check if owner of bucket has rw_access to underlying path
### Environment info
- NooBaa Version: Master
- Platform: Non Containerized

### Actual behavior
1. Currently, one can create a bucket that his owner can not access the underlying path of the bucket.

### Expected behavior
1. Currently, one can not create a bucket that his owner can not access the underlying path of the bucket.

### Steps to reproduce
```
1. mkdir /tmp/p1
2. chown 7171:7171  /tmp/p1
3. chmod 770 /tmp/p1
4. create account --name=account1 --uid=7272 --gid=7272 --new_buckets_path=/tmp/p2
5. create bucket, owner=account1 --path=/tmp/p1

```

### PR DESCRIPTION
**Pull request by shirady**
### Explain the changes
1. Add a check if the owner of the bucket has `rw_access` to the underlying path when creating or updating a bucket.
2. Move the functionality of `verify_bucket_owner` to `validate_bucket_args`  before we check the access of the underlying path (so we will have the account's details - to reduce fetching the config file again).
3. Add a new error: `InaccessibleStoragePath`.
4. Edit the existing tests to permit read and write access to the bucket owner.
5. Minor style changes and renaming in `test_nc_nsfs_bucket_cli.test.js`.

### Issues: Fixed #7883 
1. Currently, one can create a bucket that his owner can not access the underlying path of the bucket.
2. GAP - currently, we use `verify_bucket_owner` on every bucket update. It needs to be only when the owner is updated.
3. inside `validate_bucket_args` we have `data.owner_account = account._id`, we need to move it to better place.

#### Manual Tests:
1. Create the directories:
   - `path` for the bucket: `mkdir /tmp/p1`
   - `new_buckets_path` for the account: `mkdir /tmp/p2`
4. Change file owner and group of `path` directory to uid 7171 and gid 7171: `sudo chown 7171:7171  /tmp/p1`
5. Give permission: 
   - `path` for the bucket: `sudo chmod 770 /tmp/p1`
   - `new_buckets_path`  for the account: `chmod 770 /tmp/p2`
6. Create an account: `sudo node src/cmd/manage_nsfs account add --name=account1 --uid=7272 --gid=7272 --new_buckets_path=/tmp/p2`
7. Create a bucket: `sudo node src/cmd/manage_nsfs bucket add --name=account1-bucket1 --owner=account1 --path=/tmp/p1`.

#### Unit Tests:
Please run:
- `sudo npx jest test_nc_nsfs_bucket_cli.test.js` (new tests)
- `sudo node ./node_modules/.bin/_mocha src/test/unit_tests/test_nc_nsfs_cli.js` (edited tests)

- [ ] Doc added/updated
- [X] Tests added

### CODE REVIEW

File: src/test/unit_tests/jest_tests/test_nc_nsfs_bucket_cli.test.js [hunk 1]
```diff
@@ -162,6 +205,116 @@ describe('manage nsfs cli bucket flow', () => {
...
+        });
+
+        it('should fail - cli update bucket owner - owner does not have any permission to path', async () => {
+            const action = ACTIONS.UPDATE;
+            const bucket_options = { config_root, name: bucket_defaults.name, owner: account_defaults2.name};
+            await fs_utils.create_fresh_path(bucket_defaults.path);
+            await fs_utils.file_must_exist(bucket_defaults.path);
+            await set_path_permissions_and_owner(bucket_defaults.path, account_defaults2, 0o077);
+            const res = await exec_manage_cli(TYPES.BUCKET, action, bucket_options);
+            expect(JSON.parse(res.stdout).error.message).toBe(ManageCLIError.InaccessibleBucketPath.message);
```
Comment (naveenpaul1) [inline src/test/unit_tests/jest_tests/test_nc_nsfs_bucket_cli.test.js:282]: Checking error code is more common practice and safe.

File: src/cmd/manage_nsfs.js [hunk 2]
```diff
@@ -758,6 +754,13 @@ async function validate_bucket_args(data, action) {
         if (!exists) {
             throw_cli_error(ManageCLIError.InvalidStoragePath, data.path);
         }
+        const account = await verify_bucket_owner(data.bucket_owner, action);
+        if (action === ACTIONS.ADD) data.owner_account = account._id;
+        const account_fs_context = await native_fs_utils.get_fs_context(account.nsfs_account_config, data.fs_backend);
+        const accessible = await native_fs_utils.is_dir_rw_accessible(account_fs_context, data.path);
+        if (!accessible) {
+            throw_cli_error(ManageCLIError.InaccessibleBucketPath, data.path);
```
[the remark this answers is not in the document]
Comment (shirady) [reply src/cmd/manage_nsfs.js:762]: Sure, I will change, thanks.

File: src/test/unit_tests/jest_tests/test_nc_nsfs_bucket_cli.test.js [hunk 3]
```diff
@@ -162,6 +205,116 @@ describe('manage nsfs cli bucket flow', () => {
...
+
+    describe('cli update bucket', () => {
+        const config_root = path.join(tmp_fs_path, 'config_root_manage_nsfs');
+        const root_path = path.join(tmp_fs_path, 'root_path_manage_nsfs/');
+        const bucket_name = 'bucket123';
+        const bucket_storage_path = path.join(tmp_fs_path, 'root_path_manage_nsfs', bucket_name);
+        const account_name = 'account-test-update';
+        const account_name2 = 'account-test-update2';
+
+        const account_defaults = {
```
[the remark this answers is not in the document]
Comment (shirady) [reply src/test/unit_tests/jest_tests/test_nc_nsfs_bucket_cli.test.js:217]: Yes, in the `beforeEach`, we must create an account before creating a bucket when using the manage NSFS CLI.

File: src/test/unit_tests/jest_tests/test_nc_nsfs_bucket_cli.test.js (hunk 1 above)
Comment (shirady) [reply src/test/unit_tests/jest_tests/test_nc_nsfs_bucket_cli.test.js:282]: I will change, thanks.
Would you please explain about this more (or send a link) - why is it more safe?

File: src/cmd/manage_nsfs.js [hunk 4]
```diff
@@ -731,7 +727,7 @@ function get_access_keys(action, user_input) {
 async function validate_bucket_args(data, action) {
     if (action === ACTIONS.DELETE || action === ACTIONS.STATUS) {
         if (_.isUndefined(data.name)) throw_cli_error(ManageCLIError.MissingBucketNameFlag);
-    } else {
+    } else { // action === ACTIONS.ADD || action === ACTIONS.UPDATE
```
[the remark this answers is not in the document]
Comment (shirady) [reply src/cmd/manage_nsfs.js:730]: If we are in with `ACTIONS.STATUS` then we will get to the `if` condition `if (action === ACTIONS.DELETE || action === ACTIONS.STATUS)`.

File: src/test/unit_tests/jest_tests/test_nc_nsfs_bucket_cli.test.js (hunk 1 above)
Comment (naveenpaul1) [reply src/test/unit_tests/jest_tests/test_nc_nsfs_bucket_cli.test.js]: chances of updating error message is more than error code, and adding a space or dot in error message might break complete test cases. And here error code is each error have distinct error code.

Comment (shirady) [reply src/test/unit_tests/jest_tests/test_nc_nsfs_bucket_cli.test.js]: Thank you, I just don't fully understand - it is always checks for the same line, right?
For example in this line:
`expect(JSON.parse(res.stdout).error.message).toBe(ManageCLIError.InaccessibleBucketPath.message);` (before the change) it would compare this line `message: 'Account should have read & write access to the specified bucket storage path',` so by changing this line, the test should pass.

File: src/cmd/manage_nsfs.js [hunk 5]
```diff
@@ -758,6 +754,13 @@ async function validate_bucket_args(data, action) {
         if (!exists) {
             throw_cli_error(ManageCLIError.InvalidStoragePath, data.path);
         }
+        const account = await verify_bucket_owner(data.bucket_owner, action);
+        if (action === ACTIONS.ADD) data.owner_account = account._id;
```
Comment (romayalon) [inline src/cmd/manage_nsfs.js:758]: A variable assignment doesn't fit a validation function, why did you move it here?

File: src/manage_nsfs/manage_nsfs_cli_errors.js [hunk 6]
```diff
@@ -353,6 +353,11 @@ ManageCLIError.MalformedPolicy = Object.freeze({
     http_code: 400,
 });
 
+ManageCLIError.InaccessibleStoragePath = Object.freeze({
+    code: 'InaccessibleStoragePath',
+    message: 'Account should have read & write access to the specified bucket storage path',
```
Comment (romayalon) [inline src/manage_nsfs/manage_nsfs_cli_errors.js:358]:
```suggestion
    message: 'Bucket owner should have read & write access to the specified bucket storage path',
```

File: src/cmd/manage_nsfs.js (hunk 5 above)
Comment (shirady) [reply src/cmd/manage_nsfs.js:758]: I remembered that we wanted fewer calls to the validations function inside the operation, in this case, I tried to reduce the call to `verify_bucket_owner` from `add_bucket`.
In the previous, I returned the account id, and now I changed to the account since I need the `account.nsfs_account_config` and thought that by that I can reduce a call to read the account config file again.

File: src/manage_nsfs/manage_nsfs_cli_errors.js (hunk 6 above)
Comment (shirady) [reply src/manage_nsfs/manage_nsfs_cli_errors.js:358]: I will change that, thanks.

File: src/cmd/manage_nsfs.js (hunk 5 above)
Comment (romayalon) [reply src/cmd/manage_nsfs.js:768]: still, I don't think this is the place for assigning a property of the account

Comment (shirady) [reply src/cmd/manage_nsfs.js]: I removed the assignment, thanks.

**Review by romayalon — APPROVED** [review]
